src/glang/graphene_parser.py

```diff
@@ -585,19 +587,30 @@ def _function_call_outer(
         if isinstance(node, lp.UFCS_Call):
             this_arg = self.parse_expr(node.expression)
             # TODO: should there be syntax for this? It seems a bit adhoc
-            indirection_kind = this_arg.expression().underlying_indirection_kind
-            if indirection_kind == cg.Type.Kind.CONST_REF:
-                this_arg.add_parent(cg.BorrowExpression(this_arg.expression(), False))
-            elif indirection_kind == cg.Type.Kind.MUTABLE_REF:
-                this_arg.add_parent(cg.BorrowExpression(this_arg.expression(), True))
-            elif indirection_kind == cg.Type.Kind.VALUE:
-                # (&a):fn(), (&mut a):fn(), or rvalue():fn();
-                # We do not allow the 3rd option
-                possible = this_arg.expression().result_type_as_if_borrowed
-                if possible.storage_kind == cg.Type.Kind.VALUE:
-                    raise BorrowWithNoAddressError(possible.format_for_output_to_user())
-            else:
-                assert False
+            match this_arg.expression().underlying_indirection_kind:
+                case cg.Type.Kind.CONST_REF:
+                    this_arg.add_parent(
+                        cg.BorrowExpression(
+                            this_arg.expression(), cg.Type.Kind.CONST_REF
+                        )
+                    )
+                case cg.Type.Kind.MUTABLE_REF:
+                    this_arg.add_parent(
+                        cg.BorrowExpression(
+                            this_arg.expression(), cg.Type.Kind.MUTABLE_OR_CONST_REF
+                        )
+                    )
+                case cg.Type.Kind.VALUE:
+                    # (&a):fn(), (&mut a):fn(), or rvalue():fn();
+                    # We do not allow the 3rd option
+                    possible = this_arg.expression().result_type_as_if_borrowed
+                    if possible.storage_kind == cg.Type.Kind.VALUE:
+                        raise BorrowWithNoAddressError(
+                            possible.format_for_output_to_user()
+                        )
+                case _:
+                    # cg.Type.Kind.MUTABLE_OR_CONST_REF shouldn't be possible.
+                    raise AssertionError()
```

Comment:
Ooo, very modern python here.

Reply:
Here I think it's actually more readable